### ISSUE TITLE
Memory issues in DefaultResultSetHandlerFactory

### ISSUE THREAD

**Issue by ryber** [aaberg/sql2o#216]
Hello,
    We have been seeing some major GC thrashing around DefaultResultSetHandlerFactory. What we are seeing are massive cleanups around stringKey (which is a property of key). It looks like it's the same string value all the time but it's actually a new string object. This has actually caused our application to crash.

I would assume this is due to the use of StringBuilder::toString inside DefaultResultSetHandlerFactory::newResultSetHandler

StringBuilder::toString makes a new String object off of SB's internal char[]

Perhaps a better solution would be to build the key for the cache off of some kind of hashcode rather than a string?

Looking for input before attempting a solution myself.

**Comment by ryber** [comment]
![sql2ooverview](https://cloud.githubusercontent.com/assets/145359/13364389/315ed204-dc94-11e5-85ae-3390834b86cc.gif)

![sql2o](https://cloud.githubusercontent.com/assets/145359/13364392/360d2882-dc94-11e5-9548-04b8a3563c62.gif)

**Comment by ryber** [comment]
So I think I figured this out. The QuirksDetector is returning a new NoQuirks instance. So we have a new Key in the map every time the query runs even though all the other parts are the same.

The reason we get a new NoQuirks object is because QuirksProviderImpl can't figure out what our driver is since we wrap the oracle driver in ```org.apache.commons.dbcp.BasicDataSource```

Is it reasonable for the QuirksDetector to return a static final instance of NoQuirks? Should the QuirksProviderImpl dig into the BasicDataSource to find the "real" driver?

**Comment by dimzon** [comment]
Just specify proper quirks instance directly into your helper code...
27.02.2016 6:05 пользователь "Ryan Bergman" <notifications@github.com>
написал:

**Comment by ryber** [comment]
That's what I'm doing, but isn't the root code wrong for aways returning a new NoQuirks if it can't find a match? Shouldn't it be static? or change the hashcode of the key?

**Comment by dimzon** [comment]
Yes, it must be static. But other your proposal about deep provider search
against connection pool is too complex and uncommon since too many
connection pool implementations
27.02.2016 7:21 пользователь "Ryan Bergman" <notifications@github.com>
написал:

**Comment by ryber** [comment]
I've issued a PR for the 1.5 branch. I started to do the same for 1.6 but I don't think it's as good a solution there. With the creation of Quirks moved out to providers you are placing a expectation of how those are created on an interface. Those providers should return quirks in whatever way they see fit. You can't really demand that they return singletons. Additionally this is all being done so that the internal key for a internal class in a totally different class doesn't get a memory leak, which seems a dubious reason.  

Clearly the issue here is the key and the cache and not the quirks. 

Here are the options I see:
1. Change the key to use the quirk class for it's hash and not the quirk objects itself. This works fine unless you intend to differ on the map of converts that most of the quirks support (but AFAIK are never actually set). Again, they key should probably not worry about this.

2. Somehow force the quirk classes to implement some kind of name or key to use for the hash. Or force them to expose the map of converts, then the unique parts of them can be used safely.

**Comment by dimzon** [comment]
Just implement equals and hashCode by quirks implementation.
29.02.2016 2:40 пользователь "Ryan Bergman" <notifications@github.com>
написал:
